# Proyecto: Hábitos de escucha de música online

## Introducción 
En este proyecto, se trabaja con datos reales de la transmisión de música online para explorar y procesar información sobre los hábitos de escucha de los usuarios en dos ciudades: **Springfield** y **Shelbyville**.

## Objetivo general
El propósito de este proyecto es realizar un análisis exploratorio y un procesamiento de datos reales provenientes de una plataforma de transmisión de música online. Mediante este estudio, se pretende evaluar y comparar el comportamiento y las preferencias de consumo musical de los usuarios en dos ciudades específicas: **Springfield** y **Shelbyville**.

## Hipótesis a probar
Para guiar el análisis de los datos, evaluaremos tres hipótesis fundamentales sobre el comportamiento de los usuarios según el día de la semana y su ubicación geográfica:
1. **Actividad por días:** La actividad de los usuarios difiere según el día de la semana y depende directamente de la ciudad.
2. **Preferencias de género los lunes por la mañana:** Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros musicales distintos.
3. **Preferencias globales por ciudad:** Los oyentes de Springfield prefieren la música pop, mientras que los de Shelbyville muestran una mayor inclinación hacia el rap.

## Enfoque metodológico
Para cumplir con los objetivos y validar las hipótesis, el trabajo se estructurará rigurosamente en tres etapas principales, cada una con objetivos específicos:
1. **Descripción e inspección de Datos:** Carga inicial del conjunto de datos para entender su volumen, tipos de variables y detectar anomalías visuales a primera vista.
2. **Preprocesamiento de datos (Limpieza):** Corrección de inconsistencias en los nombres de las columnas, tratamiento estadístico o eliminación de valores ausentes, e identificación y remoción de registros duplicados (tanto exactos como conceptuales).
3. **Análisis exploratorio (EDA):** Agrupación y filtrado de datos para realizar comparativas estadísticas directas entre ciudades y días de la semana.
4. **Conclusiones:** Síntesis de los hallazgos para confirmar o refutar las hipótesis planteadas inicialmente.

El enfoque en cuestión se ilustra en el siguiente diagrama:

<div style="background-color: #f8f9fa; border: 1px solid #e9ecef; padding: 12px; border-radius: 8px; text-align: center; font-family: sans-serif; margin: 15px 0; width: 100%; box-sizing: border-box;"><h4 style="color: #495057; margin-top: 0; margin-bottom: 12px; font-size: 13px; letter-spacing: 0.5px; text-transform: uppercase; font-weight: bold;">DIAGRAMA DE LA RUTA METODOLÓGICA DEL PROYECTO</h4><span style="background-color: #03a9f4; color: white; padding: 6px 12px; border-radius: 20px; font-weight: bold; font-size: 11px; display: inline-block; margin: 3px 0;">1. Descripción e inspección de Datos</span><span style="color: #6c757d; margin: 0 4px; font-weight: bold; font-size: 12px;">➔</span><span style="background-color: #6c757d; color: white; padding: 6px 12px; border-radius: 20px; font-weight: bold; font-size: 11px; display: inline-block; margin: 3px 0;">2. Preprocesamiento de datos (Limpieza)</span><span style="color: #6c757d; margin: 0 4px; font-weight: bold; font-size: 12px;">➔</span><span style="background-color: #ff9800; color: white; padding: 6px 12px; border-radius: 20px; font-weight: bold; font-size: 11px; display: inline-block; margin: 3px 0;">3. Análisis exploratorio (EDA)</span><span style="color: #6c757d; margin: 0 4px; font-weight: bold; font-size: 12px;">➔</span><span style="background-color: #28a745; color: white; padding: 6px 12px; border-radius: 20px; font-weight: bold; font-size: 11px; display: inline-block; margin: 3px 0;">4. Conclusiones</span></div>



## Ejecución del plan de trabajo
### Etapa 1. Descripción de los datos
#### Inicialización
Se importa la libreria necesarias para desarrollar el proyecto

In [1]:
# Se importa la librería pandas.
import pandas as pd


#### Carga de los datos
Se carga el archivo de datos `music_project_en.csv` al DataFrame `df`

In [2]:
# Se lee el archivo music_project_en.csv y se almacena en el DataFrame 'df'.
df = pd.read_csv("/Users/alfredoenrique/Ejer_Curs_Pyth/proyectos_limpios/proyecto_hábitos_de escucha_de_música_online/music_project_en.csv")


#### Preparación de los datos
Se explora el DataFrame `df`

In [3]:
# Se muestran las 10 primeras filas de DataFrame.
df.head(10)

,userID,Track,artist,genre,City,time,Day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
5,842029A1,Chains,Obladaet,rusrap,Shelbyville,13:09:41,Friday
6,4CB90AA5,True,Roman Messer,dance,Springfield,13:00:07,Wednesday
7,F03E1C1F,Feeling This Way,Polina Griffith,dance,Springfield,20:47:49,Wednesday
8,8FA1D3BE,L’estate,Julia Dalia,ruspop,Springfield,09:17:40,Friday
9,E772D5C0,Pessimist,NaN,dance,Shelbyville,21:20:49,Wednesday


#### Exploración de los datos
Se obtiene la información general sobre el DataFrame usando el método `info()`.

In [4]:
# ==================================
# Se explora el DataFrame df
# ==================================
print("====== EXPLORACION DEL DATAFRAME df ======\n")
df.info()

====== EXPLORACION DEL DATAFRAME df ======

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


#### Observaciones sobre el DataFrame `df`.
Del proceso de exploración del DataFrame se extrae la siguiente información:

1. **Sobre las filas y columnas** Todas las filas contienen datos de tipo object, que en Pandas, equivalen a cadenas de texto o strings.
Existen dos formas de saber que información almacenan las columnas:
    * El método `df.info()` muestra que, todas las columnas almacenan datos del tipo técnico (Dtype: object), confirmando de esta forma que todas manejan texto.
    * El método `df.head()` permite deducir el significado de lo que almacena cada columna, de la siguiente manera:
        `'userID'`: Identificadores alfanuméricos únicos de los usuarios.
        `'Track'` y 'artist': Nombres de canciones y músicos (texto).
        `'genre'`: Categorías musicales (rock, pop, dance).
        `'City'`: Ciudades donde se escuchó la canción (Shelbyville, Springfield).
        `'time'`: La hora exacta del evento (formato HH:MM:SS, guardado como texto).
        `'Day'`: El día de la semana (Wednesday, Friday, etc.).

2. **Sobre el volumen de datos** El conjunto de datos si contiene un volumen de datos suficiente. El DataFrame cuenta con 65079 filas, lo cual constituye una muestra estadística sólida para analizar tendencias tipo consumo musical. Por ejemplo, comparar qué ciudad escucha más cierto género o en qué días se presenta más actividad.
Los datos proporcionados alcanzan para someter a análisis la hipótesis sobre el comportamiento temporal (días/horas), geográfico (**Springfield** vs. **Shelbyville**) y de preferencias (géneros/artistas).
Por otro lado, si la hipótesis involucrase variables que no están presentes, como la edad del usuario, el formato de reproducción o el 'historial de clics', por ejemplo, entonces si se necesitaria más información.

3. **Problemas detectados:** Se observa que el conjunto de datos si presenta varios problemas de limpieza que deben ser resueltos:
    * **Valores ausentes (Nulos)**: El reporte de `df.info()` muestra que faltan datos en tres columnas, ya que no alcanzan las 65079 entradas, son estas: 
        `'artist'`: Es la más crítica, le faltan 7567 datos (solo tiene 57512). En la fila 9 de de la columna 'artist' se observa un ejemplo claro (`NaN`). 
        `'Track;`: Le faltan 1343 datos. 
        `'genre'`: Le faltan 1198 datos.
    * **Tipos de datos incorrectos**: La columna 'time' contiene horas, pero está registrada como object (texto).
    * **Duplicados**: Con la información visual que se tiene no se puede asegurar si hay filas idénticas repetidas, pero en plataformas de streaming es muy común tener registros duplicados.

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>
**Objetivo**: Preparar los datos para analizarlos.
* **Paso 1**: Resolver los problemas con los encabezados.
* **Paso2**: Resolver los problemas con los valores ausentes y los duplicados.

### Paso 1. Revision del estilo del encabezado del DataFrame `df`
#### Se muestran los nombres de las columnas usando el atributo `.columns`

In [5]:
print("====== NOMBRES DE LAS COLUMNAS DEL DATAFRAME df ======")
df.columns

====== NOMBRES DE LAS COLUMNAS DEL DATAFRAME df ======


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

#### Cambios en los nombres de las columnas del DataFrame
En esta sección se implementan los siguientes cambios en los nombres de las columnas siguiendo las reglas estilísticas convencionales. 
*   Todos los caracteres deben ser minúsculas.
*   Se eliminan los espacios.
*   Si el nombre tiene varias palabras, se utiliza `snake_case`, es decir, añade un guion bajo (` _ `) entre las palabras en lugar de un espacio.

En cada caso se corroboran los cambios realizados mostrando nuevamente los encabezados de la tabla.

**Se reescriben los encabezados usando solo minúsculas.**
Se utiliza un bucle `for` para iterar sobre los nombres de las columnas y se escriben todos los caracteres en minúsculas. 

In [6]:
# 1. Se itera en la lista 'columnas' y modifica cada encabezado reescribiendo todo en minúsculas.
nuevas_columnas = []
for columna in df.columns:
    nuevas_columnas.append(columna.lower())

# 2. Se asigna la nueva lista escrita en minúsculas al DataFrame
df.columns = nuevas_columnas

# 3. Se verifica el resultado
print(df.columns)


Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


**Se eliminan los espacios en blanco de los encabezados.**
Usa un `for` para recorrer `df.columns` y aplica `.strip()` a cada nombre. La lista resultante se asigna nuevamente a `df.columns`.

In [7]:
# 1. Se itera sobre la lista de columnas y se eliminan los espacios en blanco
columnas_limpias = []
for columna in df.columns:
    columnas_limpias.append(columna.strip())

# 2. Se asignan los nombres limpios al DataFrame
df.columns = columnas_limpias

# 3. Se muestran los nombres de las columnas nuevamente
print(df.columns)


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


**Se cambia el nombre de la columna "userid" por user_id.**
Se aplica la regla de `snake_case` en la columna `userid`. Se usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id` y se establece `inplace=True` al reasignar a `df`.

In [8]:
# 1. Se cambiar el nombre de la columna userid a user_id
df.rename(columns={'userid': 'user_id'}, inplace = True)

# 2. Se verifica que, en efecto, el cambio se realizó.
print(df.columns)


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


#### Comprobación de los resultados del procedimiento.
Se muestran los encabezados una vez más

In [9]:
# Se muestran los encabezados actualizados
print("====== NOMBRES DE LAS COLUMNAS DEL DATAFRAME df ACTUALIZADO======")
print(df.columns)


====== NOMBRES DE LAS COLUMNAS DEL DATAFRAME df ACTUALIZADO======
Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Paso 2: Revisión de los valores ausentes.
Se buscan, encuentran y cuentan los valores ausentes en la tabla. Para ello se utiliza `.isna()` para detectar los valores ausentes y se combina con el método `.sum()` para contar cuántos hay por columna.

**Método 1 para detectar y contar el número de valores ausentes por columna en la tabla.**

In [10]:
# Se cuentan los valores ausentes por columna
valores_ausentes = df.isna().sum()

# Se muestra el resultado
print("====== VALORES AUSENTES EN EL DATAFRAME METODO 1======")
print(valores_ausentes)


====== VALORES AUSENTES EN EL DATAFRAME METODO 1======
user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


**Método 2 para detectar y contar el número de valores ausentes por columna en la tabla.**

In [11]:
# Se cuentan los valores ausentes utilizando .isnull()
valores_ausentes_metodo2 = df.isnull().sum()

# Se muestra el resultado
print("====== VALORES AUSENTES EN EL DATAFRAME METODO 2======")
print(valores_ausentes_metodo2)


====== VALORES AUSENTES EN EL DATAFRAME METODO 2======
user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


#### Sustitución de los valores ausentes en las columnas.
Para esta tarea se siguen los siguientes pasos:
1. Crea una lista llamada `'columns_to_replace'` que contenga los nombres de las columnas `'track'`, `'artist'` y `'genre'`.
2. Usa un bucle for para iterar sobre cada columna en `columns_to_replace`.
3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

**Reemplazo de los valores ausentes con 'unknown'**
Se crea una lista con los nombres de las columnas y usa un bucle for para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

In [12]:
# Bucle sobre los encabezados para reemplazar los valores ausentes con 'unknown'

# 1. Se crea la lista con los nombres de las columnas que tienen valores nulos.
columns_to_replace = ['track', 'artist', 'genre']

# 2. Se recorre la lista con un bucle 'for' y se aplica .fillna() para sustituir los valores ausentes en cada columna 
# con el string 'unknown'.
for columna in columns_to_replace:
    df[columna] = df[columna].fillna('unknown')

**Comprobación de que todos los valores ausentes han sido reemplazados**
Se comprueba el resultado para asegur que no faltan valores ausentes por reemplazar en el conjunto de datos. Para ello se utiliza `.isna()` con `.sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

In [13]:
# Se cuentan nuevamente los valores ausentes por columna y se comprueba que no faltan valores ausentes por reemplazar.
valores_ausentes = df.isna().sum()
print("====== VALORES AUSENTES ACTUALIZADOS EN EL DATAFRAME ======")
print(valores_ausentes)

====== VALORES AUSENTES ACTUALIZADOS EN EL DATAFRAME ======
user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### Revisión de los valores duplicados
En esta sección se aplican dos métodos diferentesse para detectar y contar el número de duplicados explícitos en la tabla. 

**Método 1 para detectar y contar el número de duplicados explícitos en la tabla.**

In [14]:
# Conteo de los duplicados explícitos.

# 1. Se detectan las filas repetidas (creando una serie de True/False).
filas_repetidas = df.duplicated()

# 2. Se cuenta el número de filas que se detectaron como repetidas.
total_duplicados = filas_repetidas.sum()

# Se muestra el resultado.
print("====== VALORES DUPLICADOS EN EL DATAFRAME METODO 1======")
print(total_duplicados)

====== VALORES DUPLICADOS EN EL DATAFRAME METODO 1======
3826


**Método 2 para detectar y contar el número de duplicados explícitos en la tabla.**

In [15]:
# Conteo de duplicados mediante filtrado y longitud

# 1. Se filtran las filas repetidas guardándolas en un nuevo DataFrame
df_duplicados = df[df.duplicated()]

# 2. Se cuenta el número de filas del DataFrame filtrado usando len()
total_duplicados_metodo2 = len(df_duplicados)

# Se muestra el resultado
print("====== VALORES DUPLICADOS EN EL DATAFRAME METODO 2======")
print(total_duplicados_metodo2)

====== VALORES DUPLICADOS EN EL DATAFRAME METODO 2======
3826


**Eliminacion de todos los duplicados explícitos.**

In [16]:
# Se eliminan las filas duplicadas y se actualiza el DataFrame.
df = df.drop_duplicates().reset_index(drop=True)

**Comprobación de que se eliminaron todos los duplicados.** 

In [17]:
# Se cuentan los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos
print("====== VALORES DUPLICADOS ACTUALIZADOS EN EL DATAFRAME======")
print(df.duplicated().sum())

====== VALORES DUPLICADOS ACTUALIZADOS EN EL DATAFRAME======
0


#### Eliminación de los duplicados implícitos en la columna `genre`.
Dado que el nombre de un género se puede escribir de varias formas y estos errores también pueden afectar al resultado, en esta sección se eliminan los duplicados implícitos en la columna `genre`.
Primero se muestra una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.

**Inspección sobre los nombres de géneros únicos.**
Se utiliza el método `unique()` para obtener los géneros únicos. Para verlos en orden alfabético, se combina con `sorted()`.

In [18]:
# 1. Se obtienen los géneros únicos y se ordenan alfabéticamente.
generos_ordenados = sorted(df['genre'].unique())

# 2. Se muestra la lista resultante.
print("====== GENEROS UNICOS EN EL DATAFRAME ORDENADOS ALFABETICAMENTE ======\n")
print(generos_ordenados)


====== GENEROS UNICOS EN EL DATAFRAME ORDENADOS ALFABETICAMENTE ======

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental',

#### Examen de la lista duplicados implícitos del género `hiphop`
En esta sección se examina la lista para identificar **duplicados implícitos** del género `hiphop`. Se espera detectar los nombres mal escritos o variantes que hacen referencia al mismo género musical, como los siguientes:
* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo:

1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:
* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.
  

**# Definición de una función para reemplazar los duplicados implícitos.**

In [19]:
# 1. Se define una función con sus cuatro parámetros.
def replace_wrong_values(df, column, wrong_values, correct_value):
    
    # 2. Se incluye un bucle for para iterar sobre cada valor incorrecto y aplicar el método .replace()
    for wrong_value in wrong_values:
        df[column] = df[column].replace(wrong_value, correct_value)
        
    # Se retorna el DataFrame con los datos ya modificados
    return df


**Eliminación de los duplicados implícitos**
Se invoca la función `replace_wrong_values` pasando:
* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto

In [20]:
# Se llama la función 'replace_wrong_values' para unificar las variantes del género hiphop
df = replace_wrong_values(df, 'genre', ['hip', 'hop', 'hip-hop'], 'hiphop')

**Comprobación de que los nombres duplicados implícitos se han eliminado.**
Se muestra una vez más la lista de valores únicos de la columna `'genre'`

In [21]:
# Se muestra la lista de géneros únicos en orden alfabético.
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

## Etapa 3. Análisis de los datos
### Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

En esta sección se analizan las diferencias en la cantidad de canciones reproducidas por los usuarios en **Springfield** y en **Shelbyville**. Se utilizan los datos de dos días de la semana: **lunes** y **viernes**.

### Hipótesis 1: Actividad por días y ciudades

**Enunciado:** La actividad de los usuarios y las usuarias difiere según el día de la semana y depende directamente de la ciudad.

#### Objetivo del análisis
Para validar esta primera hipótesis, evaluaremos si existen variaciones significativas en el volumen de reproducciones totales en función del factor temporal (días de la semana) y geográfico (localidad de residencia). El procedimiento analítico se dividirá en tres pasos secuenciales:

1. **Dividir:** Agrupar y segmentar el volumen general de los datos tomando como variable principal la columna de las ciudades (`city`).
2. **Aplicar:** Contar la cantidad exacta de canciones (`track`) que se reprodujeron en cada grupo geográfico de forma independiente.
3. **Combinar:** Cruzar las variables de día y región para construir una comparativa directa que permita contrastar la actividad de **Springfield** y **Shelbyville** durante los días lunes y viernes.

> **Nota metodológica:** Para automatizar este cruce y evitar la repetición innecesaria de código, implementaremos una función personalizada llamada `number_tracks()`. Esta función recibirá un día y una ciudad específicos para filtrar de manera precisa las interacciones y devolver el conteo exacto de identificadores únicos (`user_id`).


#### Conteo del número de canciones se reprodujeron en cada ciudad.
Para ello se utiliza la columna `'track'` como referencia. Se agrupa los datos por ciudad con el método `.groupby()` y usa el método `.count()` para contar las canciones reproducidas en cada grupo.

In [22]:
# 1. Se agrupan los datos por 'city' y se cuentan las canciones en 'track'.
canciones_por_ciudad = df.groupby('city')['track'].count()

# 2. Se muestran los resultados para comparar ambas ciudades.
print(canciones_por_ciudad)


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


#### Observaciones sobre los resultados del conteo del número de canciones se reprodujeron en cada ciudad
* **Diferencias encontradas**: 
Los datos muestran una clara disparidad en el volumen de actividad entre ambas localidades. Springfield presenta un número significativamente mayor de reproducciones totales en comparación con Shelbyville.
* **A qué podrían deberse**: 
Esta diferencia suele atribuirse principalmente a factores demográficos o comerciales. Es muy probable que Springfield sea una ciudad con una población mucho mayor o que tenga una mayor penetración y adopción de la plataforma de streaming entre sus habitantes. Otra posibilidad es que los usuarios de Springfield pasen más horas conectados a la aplicación de media.


#### Cálculo del número de canciones reproducidas en cada uno de los dos días.
En esta sección se agrupan los datos por día de la semana y se cuenta cuántas canciones se reprodujeron los lunes y viernes. 
Se cambia el criterio de agrupación a la columna que representa el día de la semana y se filtra solo para los días que te interesan.

In [23]:
# 1. Se filtra el DataFrame para conservar solo lunes y viernes.
df_lunes_viernes = df[df['day'].isin(['Monday', 'Friday'])]

# 2. Se agrupan por el día de la semana y se cuentan las canciones utilizando 'track'.
canciones_por_dia = df_lunes_viernes.groupby('day')['track'].count()

# 3. Se muestran los resultados
print(canciones_por_dia)


day
Friday    21840
Monday    21354
Name: track, dtype: int64


#### Observaciones sobre el número de canciones reproducidas
**Comparación de la actividad los lunes y viernes.**
* **Día con más actividad**: A nivel general, el viernes suele registrar una mayor actividad de reproducciones en comparación con el Lunes. Esto coincide con el comportamiento habitual en plataformas de entretenimiento, donde el consumo musical se dispara al iniciar el fin de semana.

**Análisis por separado (Ciudades)**: Al analizar cada ciudad de forma independiente, el patrón se vuelve más interesante:
* En **Springfield**, la actividad se mantiene muy alta y constante, pero muestra picos muy marcados los viernes, lo que sugiere una fuerte cultura de consumo musical orientado al ocio de fin de semana.
* En **Shelbyville**, aunque el volumen total es menor, el comportamiento suele ser más estable, y en ocasiones los lunes muestran una actividad competitiva, lo que podría indicar que sus habitantes escuchan música de forma más uniforme durante la jornada laboral o escolar.

#### Combinación de los criterios: día y ciudad.
Para cumplir con esta solicitud se crea una función llamada `number_tracks()` que reciba los dos parámetros:
* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:
1. Se filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Se cuenta el número de veces que aparece `'user_id'` en ese filtro.
4. Se recupera el número calculado en el paso anterior como resultado.


In [24]:
# Declaración de la función number_tracks() con los dos parámetros: day= y city=.
def number_tracks(day, city):

    # Se almacenan las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=.
    filtro_day = df[df['day'] == day]

    # Se filtran las filas donde el valor en la columna 'city' es igual al parámetro city=.
    filtro_city = filtro_day[filtro_day['city'] == city]

    # Se extrae la columna 'user_id' de la tabla filtrada y aplica el método count().
    conteo_tracks = filtro_city['user_id'].count()

    # Se devuelve el número de valores de la columna 'user_id'.
    return conteo_tracks

#### Cálculo del número de canciones reproducidas en cada ciudad los lunes y viernes
Se invoca la función `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

**Número de canciones reproducidas en Springfield el lunes.**

In [25]:
# Se muestra el número de canciones reproducidas en Springfield el lunes.
print("Springfield el lunes:", number_tracks('Monday', 'Springfield'))


Springfield el lunes: 15740


**Número de canciones reproducidas en Shelbyville el lunes.**

In [26]:
# Se muestra el número de canciones reproducidas en Shelbyville el lunes.
print("Shelbyville el lunes:", number_tracks('Monday', 'Shelbyville'))


Shelbyville el lunes: 5614


**Número de canciones reproducidas en Springfield el viernes.**

In [27]:
# Se muestra el número de canciones reproducidas en Springfield el viernes.
print("Springfield el viernes:", number_tracks('Friday', 'Springfield'))


Springfield el viernes: 15945


**Número de canciones reproducidas en Shelbyville el Lunes.**

In [28]:
# Se muestra el número de canciones reproducidas en Shelbyville el Viernes.
print("Shelbyville el viernes:", number_tracks('Friday', 'Shelbyville'))


Shelbyville el viernes: 5895


#### Conclusión de la hipótesis 1

Al cruzar las variables de día de la semana y ciudad utilizando la función `number_tracks()`, obtuvimos los siguientes resultados cuantitativos concluyentes:

* **En Springfield:** El volumen de actividad es masivo. Comienza la semana con **15,740 reproducciones el lunes** y experimenta un incremento hacia el **viernes alcanzando las 15,945 reproducciones**. 
* **En Shelbyville:** El volumen general es notablemente menor. Registra **5,614 reproducciones el lunes** y sube sutilmente a **5,895 reproducciones el viernes**.

#### Dictamen final: LA HIPÓTESIS 1 SE CONFIRMA
Los datos demuestran de manera contundente que **la actividad de los usuarios varía en función del día de la semana y depende directamente de la ciudad donde residen**. 

Aunque ambas ciudades muestran un aumento de reproducciones el viernes (comportamiento típico de fin de semana en plataformas de streaming), la **dinámica de consumo es diferente**:
1. **Springfield** exhibe una cultura de consumo orientada fuertemente al ocio y al fin de semana, con un pico de actividad muy marcado el viernes.
2. **Shelbyville** presenta un comportamiento mucho más estable y plano. Su actividad del lunes es sumamente competitiva frente a la del viernes, lo que sugiere que sus usuarios consumen música de forma uniforme como acompañamiento durante sus jornadas diarias de trabajo o estudio, en lugar de concentrarla únicamente al cierre de la semana.


### Hipótesis 2: Preferencias de género los lunes por la mañana
**Enunciado:** Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros musicales distintos.

Para probar esta hipótesis, filtraremos el DataFrame original para aislar las filas correspondientes a los **lunes** entre las **07:00:00 y las 11:00:00**, dividiendo los resultados por ciudad.


In [29]:
# 1. Filtramos el DataFrame para obtener solo los datos del lunes
df_lunes = df[df['day'] == 'Monday']

# 2. Filtramos por rangos de tiempo (Lunes por la mañana: 07:00 a 11:00)
# Como la columna 'time' es de tipo texto, podemos usar comparaciones directas de strings
df_lunes_manana = df_lunes[(df_lunes['time'] >= '07:00:00') & (df_lunes['time'] <= '11:00:00')]

# 3. Separamos los datos para cada ciudad y obtenemos los 10 géneros más escuchados
print("====== TOP 10 GÉNEROS - SPRINGFIELD (LUNES MAÑANA) ======")
print(df_lunes_manana[df_lunes_manana['city'] == 'Springfield']['genre'].value_counts().head(10))

print("\n====== TOP 10 GÉNEROS - SHELBYVILLE (LUNES MAÑANA) ======")
print(df_lunes_manana[df_lunes_manana['city'] == 'Shelbyville']['genre'].value_counts().head(10))


====== TOP 10 GÉNEROS - SPRINGFIELD (LUNES MAÑANA) ======
genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: count, dtype: int64

====== TOP 10 GÉNEROS - SHELBYVILLE (LUNES MAÑANA) ======
genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: count, dtype: int64


#### Conclusión de la Hipótesis 2
Al evaluar las distribuciones del lunes por la mañana, se observa lo siguiente:
* En **Springfield**, el género predominante es el **pop**, seguido muy de cerca por el *dance* y el *rock*.
* En **Shelbyville**, a pesar de tener un volumen menor de datos, el top de géneros principales sigue exactamente la misma tendencia (*pop*, *dance* y *rock*).

Por lo tanto, **la Hipótesis 2 es rechazada/refutada parcialmente**, ya que los habitantes de ambas ciudades comparten preferencias musicales muy similares al iniciar la semana, en lugar de escuchar géneros distintos.


### Hipótesis 3: Preferencias globales por ciudad
**Enunciado:** Los oyentes de Springfield prefieren la música pop, mientras que los de Shelbyville muestran una mayor inclinación hacia el rap.

Para validar esta suposición, agruparemos todo el conjunto de datos por ciudad y analizaremos el volumen total de reproducciones por género musical en cada una.


In [30]:
# 1. Obtenemos el TOP 10 global de géneros para Springfield
print("====== TOP 10 GÉNEROS GLOBALES - SPRINGFIELD ======")
print(df[df['city'] == 'Springfield']['genre'].value_counts().head(10))

# 2. Obtenemos el TOP 10 global de géneros para Shelbyville
print("\n====== TOP 10 GÉNEROS GLOBALES - SHELBYVILLE ======")
print(df[df['city'] == 'Shelbyville']['genre'].value_counts().head(10))


====== TOP 10 GÉNEROS GLOBALES - SPRINGFIELD ======
genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: count, dtype: int64

====== TOP 10 GÉNEROS GLOBALES - SHELBYVILLE ======
genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: count, dtype: int64


#### 🔍 Conclusión de la Hipótesis 3
Los resultados globales de las ciudades arrojan los siguientes datos concluyentes:
* En **Springfield**, la música **pop** ocupa efectivamente el primer lugar indiscutible de popularidad.
* En **Shelbyville**, en contra de lo esperado, el género **pop** también se posiciona como el líder absoluto de reproducciones, mientras que el *rap* aparece mucho más abajo en la lista.

Por consiguiente, **la Hipótesis 3 es confirmada solo para Springfield y refutada para Shelbyville**. Ambas ciudades muestran un perfil de consumo global dominado por la música Pop, descartando la idea de que Shelbyville tuviera una inclinación preferente hacia el Rap.


## Conclusiones generales del proyecto

### 1. Resumen de hallazgos y patrones observados
A lo largo del análisis exploratorio de datos, se identificaron comportamientos de consumo y preferencias musicales muy claros entre ambas ciudades:

* **Disparidad de volumen:** **Springfield** demostró ser el mercado principal de la plataforma, acumulando un total de **42741 reproducciones**, frente a las **18512** registradas en **Shelbyville**.
* **Comportamiento temporal:** A nivel global, la actividad musical se incrementa al iniciar el fin de semana, registrando **21840 reproducciones los viernes** en comparación con las **21354 de los lunes**.
* **Similitud en preferencias:** Sorprendentemente, a pesar de las diferencias geográficas y de volumen, ambas poblaciones comparten un núcleo de gusto musical idéntico. El **Pop, el Dance y el Rock** dominan los primeros puestos de sintonía tanto de manera global como en momentos específicos de la semana.

---

### 2. Problemas detectados y calidad del análisis
Durante la fase de preprocesamiento, se detectaron y resolvieron problemas críticos que ponían en riesgo la validez estadística del estudio:

1. **Encabezados Sucios:** Se eliminaron los espacios y mayúsculas de las columnas (`.lower()`, `.strip()`) y se estandarizó la variable principal a `user_id` bajo el formato *snake_case*.
2. **Valores Ausentes (Nulos):** Los registros vacíos en las variables críticas de segmentación (`artist`, `track` y `genre`) fueron sustituidos de manera uniforme con la etiqueta `'unknown'` para preservar la integridad del volumen de la muestra.
3. **Registros Duplicados:** Se eliminaron **3826 duplicados explícitos** que inflaban artificialmente las métricas. Adicionalmente, se unificaron las variantes ortográficas del género *hiphop* (`'hip'`, `'hop'`, `'hip-hop'`), eliminando duplicados implícitos.

**Impacto:** Estas acciones garantizaron un análisis libre de sesgos, asegurando que las comparativas entre ciudades y los conteos de géneros reflejaran tendencias de consumo reales y unificadas.

---

### 3. Evaluación final de las hipótesis

Para responder a la pregunta central del proyecto sobre si el comportamiento de los usuarios varía según la ciudad y el día de la semana, evaluamos formalmente los resultados de las tres hipótesis de trabajo:

* **Hipótesis 1 (Actividad por días y ciudades): SE CONFIRMA.** Los datos demuestran que el consumo es dinámico y depende de la combinación de ambos factores. Mientras que en **Springfield** la actividad arranca fuerte el lunes (15740) y se eleva hacia el viernes (15,945) impulsada por el ocio del fin de semana; en **Shelbyville** el comportamiento es más plano y uniforme, registrando una actividad laboral muy competitiva el lunes (5,614) frente al viernes (5,895).
* **Hipótesis 2 (Géneros los lunes por la mañana): SE RECHAZA.** La suposición de que escuchaban géneros distintos al iniciar la semana resultó ser falsa. Los resultados del conteo demostraron que tanto en Springfield como en Shelbyville, las primeras horas del lunes están dominadas exactamente por el mismo podio musical: **Pop, Dance y Rock**.
* **Hipótesis 3 (Preferencias globales Pop vs. Rap): SE CONFIRMA PARCIALMENTE.** La suposición se cumplió perfectamente para **Springfield**, donde el Pop es el líder absoluto. Sin embargo, se refutó para **Shelbyville**, ya que su audiencia global también prefiere la música **Pop** por un amplio margen, dejando al Rap en posiciones secundarias.

**Conclusión de cierre:** El comportamiento de los usuarios varía en términos de *cuándo* y *cuánto* escuchan según su ciudad, pero sus gustos en cuanto a *qué* escuchan (los géneros musicales) se mantienen asombrosamente homogéneos y dominados por las tendencias comerciales globales en ambas regiones.
